cmd/finch/gen_docs.go

```diff
@@ -0,0 +1,231 @@
+	genDocsGenerateCommand := &cobra.Command{
+		Use:    "generate",
+		Args:   cobra.NoArgs,
+		Hidden: true,
+		Short:  "Generate Finch docs",
+		RunE:   newGenDocsGenerateAction(rootCmd, logger, fs, deps).runAdapter,
+	}
+
+	genDocsGenerateCommand.Flags().StringP("path", "p", "", "Doc output directory")
+	// Ignore error since we check if the flag is set anyway
+	_ = genDocsGenerateCommand.MarkFlagRequired("path")
+
+	return genDocsGenerateCommand
+}
+
+type genDocsAction struct {
+	rootCmd *cobra.Command
+	logger  flog.Logger
+	fs      afero.Fs
+	deps    GenDocsSystemDeps
+}
+
+func newGenDocsGenerateAction(
+	rootCmd *cobra.Command,
+	logger flog.Logger,
+	fs afero.Fs,
+	deps GenDocsSystemDeps,
+) *genDocsAction {
+	return &genDocsAction{
+		rootCmd: rootCmd,
+		logger:  logger,
+		deps:    deps,
+		fs:      fs,
+	}
+}
+
+func (gd *genDocsAction) runAdapter(cmd *cobra.Command, _ []string) error {
+	path, err := cmd.Flags().GetString("path")
+	if err != nil {
+		return fmt.Errorf("failed to get required parameter 'path': %w", err)
+	}
+	return gd.run(path)
+}
+
+func (gd *genDocsAction) run(outDir string) error {
+	return gd.captureHelpOutput(gd.rootCmd, outDir)
+}
+
+func (gd *genDocsAction) captureHelpOutput(cmd *cobra.Command, outDir string) error {
+	for _, c := range cmd.Commands() {
+		if err := gd.captureHelpOutput(c, outDir); err != nil {
+			return err
+		}
+	}
+
+	if !cmd.Runnable() || cmd.Hidden {
+		return nil
+	}
+
+	gd.logger.Infof("Creating doc for command: %s", cmd.CommandPath())
+
+	baseName := strings.ReplaceAll(cmd.CommandPath(), " ", "_") + ".md"
+	fileName := filepath.Clean(filepath.Join(outDir, baseName))
+	f, err := gd.fs.OpenFile(fileName, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0o666)
+	if err != nil {
+		return fmt.Errorf("error while creating out file %q: %w", fileName, err)
+	}
+
+	// redirect Stdout to pipe
+	rescueStdout := gd.deps.Stdout()
+	r, w, err := gd.deps.Pipe()
+	if err != nil {
+		return fmt.Errorf("error while creating pipe to capture stdout: %w", err)
+	}
+	gd.deps.SetStdout(w)
+
+	rootCmd := cmd.Root()
+
+	// everything except the initial `finch`
+	args := strings.Split(cmd.CommandPath(), " ")[1:]
+	args = append(args, "--help")
+	rootCmd.SetArgs(args)
+	// rootCmd.SetOutput() would work for all "default finch" commands, but doesn't work
+	// for the nerdctl commands. Getting it to work would remove the need to capture all Stdout.
+	if err := rootCmd.Execute(); err != nil {
+		// This is pretty much impossible because cobra checks if --help is set and if it is it doesn't
+		// actually run the command.
+		// https://github.com/spf13/cobra/blob/main/command.go#L1096-L1099
+		return fmt.Errorf("error while executing command (args=%v): %w", args, err)
+	}
+	gd.deps.SetStdout(rescueStdout)
+
+	_ = w.Close()
+	helpText, err := io.ReadAll(r)
+	if err != nil {
+		return fmt.Errorf("error while reading stdout from pipe: %w", err)
+	}
+
+	seeAlso := ""
+	if cmd.HasParent() && cmd.Parent().Runnable() {
```

Comment:
nit: can we add an example for this branch in code, i.e when the parent command is runnable.

Reply:
Sure, a contrived example is that if you could run `finch images` as well as `finch images delete`, finch images is the parent command and finch images delete is the child command, so both commands would get their own page.

The technical explanation is the a `cobra.Command` is runnable if it has a `Run` or `RunE` method defined.

Reply:
Added comment in latest commit

Reply:
Cool thanks for the example.